### DISCUSSION TITLE
NcJump: a jumping project

### DISCUSSION THREAD

**Discussion by Fahien** [nCine/nCine#10] [Show and tell]
Last month I decided to work on [ncJump](https://github.com/Fahien/ncJump), a project with a new engine: [nCine](https://github.com/nCine/nCine), an open-source C++ framework for 2D games developed by @encelo.

Some time ago, a bunch of members of the [GameLoop](https://gameloop.it) community followed [CS50's introduction to game development](https://cs50.harvard.edu/games/2018/), and I though the sequence of games presented in this course could give me a good insight on what kind of game to create.

A couple of developers have already made or started to make the first few games listed by that course. There is [a Pong clone](https://github.com/nCine/ncPong) made by @encelo, a Flappy Bird clone made by @Vasile-Peste, and a Breakout clone under development by [@mat3](). I opted for a Super Mario Bros clone.

The idea was to follow the steps of [CS50's lecture 4](https://cs50.harvard.edu/games/2018/weeks/4/) by using nCine instead of [Love2D](https://love2d.org/).

Some steps are required in order to start developing a game, but luckily everything you need to know is documented on the [nCine download web-page](https://ncine.github.io/download/). I installed nCine, cloned [ncTemplate](https://github.com/nCine/ncTemplate), and started working on top of it.

## Writing down some code

The main thing to consider from a software design perspective is that I renamed `MyEventHandler` to `JumpHandler` and introduced a `Game` class. While `JumpHandler` acts as a bridge between the game and the engine, `Game` is the nucleus of the project and it is responsible for the lifetimes of its objects and the logic of the application.

Then I proceeded implementing an `Entity` class following the idea of an [Entity component system](https://en.wikipedia.org/wiki/Entity_component_system). The entity can have components like a graphics sprite, a finite state machine, or a body for physics and collisions.

A GUI is mandatory to quickly iterate on new elements as they are added to the game, both to have a visual feedback when debugging is needed and to tweak all sort of values and configurations at run-time. For this, I introduced an `Editor` class, which makes an heavy use of @ocornut's [Dear ImGui](https://github.com/ocornut/imgui), also provided directly by nCine.

[Box2D](https://box2d.org/) is my choice for 2D physics and collisions. It is made by @erincatto and it is very easy to use.

I spent the last weeks implementing some key Entity states: idle, move, jump, and fall. In my opinion they are the foundations of the whole gameplay and I believe the result is good enough to move on to the next steps, but for those you will need to wait for the next devlog, so stay tuned!

https://user-images.githubusercontent.com/6058008/103419322-d8d3ce00-4b92-11eb-826b-971a4811cffd.mp4

**Comment by encelo** [comment]
Great topic, thank you @Fahien.
I will soon have a look at the sources as I'm curious about a couple things you mention, like the ECS and the Box2D integration. 😉 

Keep us posted about the progress of this project! 👍🏻

**Comment by Fahien** [comment]
Happy new year with a new devlog about the progress on [ncJump](https://github.com/Fahien/ncJump). I spent some time on refactoring the Entity source code that were becoming too monolithic, and improving tileset and tilemap by using entities instead of raw sprites for both tile prototypes and concrete tiles. So let us jump into the details.

# Everything is entity

An important consideration is that everything is an [entity](https://en.wikipedia.org/wiki/Entity_component_system). Characters are entities, tiles are entities, triggers are entities, and so on. So how do you distinguish between them? Every entity can have multiple components, whether it is an image or a [state machine](https://gameprogrammingpatterns.com/state.html) or a physics body, therefore different set of components will determine all the different kinds of entities which will populate a game scene. In this devlog we are going to explore the various components needed for the second iteration of the ncJump development.

| component | optional |
|-|-|
| Transform | no |
| Graphics | yes |
| Physics | yes |
| State | yes |

### Transform

The [transform](https://www.gdcvault.com/play/1017652/Math-for-Game-Programmers-Matrix) component is simply a wrapper around a `ncine::SceneNode`. Everything else is supposed to be related to this node, so if we want to move the overall entity, we can just update the node of the transform component. It is worth mentioning that we use a unique pointer to store the node on the heap to make sure its memory would not move. This is needed as the nCine scene graph works with pointers.

### Physics

The physics component as well as acting as an abstraction layer between the game and [Box2D](https://box2d.org/), it holds values like the maximum velocity, jumping strength, and so on. Most importantly it keeps track of a handle to a [b2Body](https://box2d.org/documentation/classb2_body.html) which is the mean to query the current physical state of the entity.

### Graphics and State

At the base level we have a graphics component with one single image or animation, and a state component with one single state which never changes. These are possibly the most common among the entities of a game.

Now we need to consider the concept of a moving character with states like idle, move, jump, and fall. This set of states is also very common so we implement it directly in C++. The result is the following class hierarchy.

| Subclass | states |
|-|-|
| SingleState | IDLE |
| CharacterState | IDLE, MOVE, JUMP, FALL |

It follows that a `CharacterState` class is closely coupled to its related graphics component, which we define as `CharacterGraphics`, so the graphics components class hierarchy is quite similar to the previous one.

| Subclass | graphics |
|-|-|
| SingleGraphics | IDLE |
| CharacterGraphics | IDLE, MOVE, JUMP, FALL |

As a final consideration, while a state component should be unique for each entity, a graphics component could be shared between multiple entities so to avoid duplicating these kinds of heavy objects.

https://user-images.githubusercontent.com/6058008/103491166-c02d1780-4e21-11eb-9327-09faaa151583.mp4

**Reply by encelo** [reply]
It's getting more and more interesting with time! 😱
Thanks again for sharing the progress here. 🙏🏻

**Comment by Fahien** [comment]
Here I am with another devlog. This week I introduced a **camera component**, and **serialization** to load and save to file the state of the game. While the former was pretty easy and straightforward to implement, the latter required introducing a new dependency and various scattered refactorings to pay some _technical debt_ accumulated from the previous development iterations.

## The smooth

That was easy. The camera component has two main attributes: the scene root node and the target node. The scene root node is needed because we need to simulate the camera, which means that what we actually transform is the scene root node. In other words, to move the camera in on direction, we translate the scene root node to the opposite of such a direction. The target node is the one the camera should follow. As the target node moves around the scene, the camera is updated to match its coordinates.

The quickest thing to implement a following camera is to just write down something like `camera.position = target.position`, and while it works it is not the best we could do. A better and nicer approach, which might also prevent motion sickness, is to implement a camera with a smooth movement. This can be done by linearly interpolating the camera and the target position by a small factor at every update of the game.

```cpp
void Camera::update() {
  float smooth_factor = 0.125f;
  Vec2f smoothed_position = lerp(this->position, target->position, smooth_factor);
  this->position = smoothed_position;
}
```

https://user-images.githubusercontent.com/6058008/104129576-b731ce00-536c-11eb-9ec5-19f99cffebfb.mp4

## The file

Conversely, this feature required a bit more work. I started by adding [@nlohmann](https://github.com/nlohmann)'s [json](https://github.com/nlohmann/json) as a new dependency of the project for reading and writing to json files. I find it very easy to use in addition to the fact that I already used it for previous projects, like [Jamspot: Sokoban](https://github.com/GameLoop-it/jamspot/commit/931bba660fc9f855cbdb48b76dac825c57cf12c6). The second step was implementing functions to save certain objects to and load them from file.

```cpp
void to_json(json& j, const T& t);
void from_json(const json& j, T& t);
```

The trickiest part was to make sure that all the _serializable_ classes of objects could be _default constructed_. That means avoiding references as class members and preferring pointers, hence allowing semi-initialized objects while making sure they stay in this state for the shortest period as possible to avoid logic errors or even crashes.

Once all this has been taken care of, I was able to edit the game configuration, the tileset, and the tilemap at runtime, close the game, and re-open it to find everything as I set it before. That is key to work on the levels of the game, which is going to be the next iteration of the [ncJump](https://github.com/Fahien/ncJump) development.

![persistence](https://user-images.githubusercontent.com/6058008/104129583-c1ec6300-536c-11eb-890f-5fe3b3cfecc9.png)

**Comment by Fahien** [comment]
Welcome to this new devlog on the development of [ncJump](https://github.com/Fahien/ncJump). Previously I wrote about improvements on movement on the camera and serialization of the state of the game, and I concluded mentioning about being able to start crafting actual levels. Some bug fixing and generic improvements were needed on various aspects of my code, but I managed to shape something interesting that, although it may seem incomplete, it is a convincing baseline.

https://user-images.githubusercontent.com/6058008/104849095-48aebc00-58e8-11eb-9571-50f2e06a82e6.mp4

## Jump higher

From a gameplay perspective, the player gains confidence with the elements of interactivity of the game. It first moves, then jumps, then jumps higher. The implementation is straightforward thanks to [Box2D](https://box2d.org/). If the _jump button_ is still down, keep applying a small force upwards. Keep in mind that force should be weaker than gravity, otherwise the character would take off into space.

```cpp
void can_jump_higher(Input& input, Entity& entity) {
  if (input.jump.down) {
    auto force = b2Vec2(0.0f, small_amount);
    entity.physics->body->ApplyForceToCenter(force, true);
  }
}
```

## Resizing the tilemap

This was less straightforward. I am starting to believe the more something is boring the more is difficult to implement. Resizing is not trivial, many things should be considered. For simplicity here I will focus on the concrete tiles grid only. At the beginning I used a simple vector to store all concrete tiles of the map and with a simple formula (`y + x * height`) I was able to index a cell of the grid. Then I realized that resizing this vector was not convenient as the order of the tiles in the cells would not be preserved.

As a solution, I changed the structure from a simple vector to a vector of vectors, so we could say a list of columns. That simplified indexing as you could just use the [subscript operator](https://docs.microsoft.com/en-us/cpp/cpp/subscript-operator) to access the right tile (`tiles[x][y]`).

It also simplified resizing. If you want to change the width of the tilemap, you just modify the number of columns by calling `tiles.resize(new_width)`. As you can imagine, if the width shrinks, you lose those columns, while if the width grows, you get new columns with a _default_ concrete tile. I can not overstate the [importance of default values](https://www.codemotion.com/magazine/dev-hub/gamedev/john-romero-25-years-of-doom/). Changing the height is slightly different, as you need to resize all the columns.

```cpp
for (uint32_t i = 0; i < width; ++i) {
  tiles[i].resize(new_height);
}
```

I hope you enjoyed this devlog. Stay tuned for the next one as I will discuss about dynamic entities!

**Comment by Fahien** [comment]
Back with a new devlog! This time I am going to talk about dynamic entities, and how everything is now destructible!

## The easy stuff

While a static entity is not affected by physical forces thus does not move, a dynamic entity is affected by forces, so it can fall and can be pushed and pulled around. Thanks again to [Box2D](https://box2d.org/), implementing this was just a matter of setting the type of physics bodies to either `b2_dynamicBody` or `b2_staticBody`.

https://user-images.githubusercontent.com/6058008/106210904-075fbb80-61c8-11eb-93df-d9479797332a.mp4

Another easy but important thing to add was a proper background to give to players' eyes something more interesting to see than a boring solid color. For the moment, I have just added a static sprite, but I guess it would be nice to implement a fancy [parallax scrolling effect](https://en.wikipedia.org/wiki/Parallax_scrolling) in the future.

![github](https://user-images.githubusercontent.com/6058008/106210940-17779b00-61c8-11eb-8adb-76ff0b62a29b.png)

At this point, I realized that something was missing. Tiles and entities could be added and positioned freely in the scene and, as all normal people that like symmetry, I started wanting to destroy them, not by interacting through the editor, but as Mario does by punching (or hitting with his head?)!

## Appetite for Destruction

Well, not every tile should be destructible, but some of them certainly could, and I supposed it could be a good idea to give this ability to the player.

I started by introducing a flag to determine whether a tile is destructible or not. Then I implemented a *destruction* listener, which is a [Box2D contact listener](https://box2d.org/documentation/classb2_contact_listener.html), responsible of checking impulses generated from collisions between entities. When this impulse is big enough, both flags of the entities involved in a collision are checked and, if destructible, added to a list. This list is processed later when the actual destruction happens. It is important to note that should not be done in the previous step if you want to avoid errors while Box2D is still processing its update.

After tweaking the threshold for the impulse triggering destructions, last thing to do was to increase manually impulses when the collision involved a character punching. In technical words, when one of the entities is a *character* and the normal of the collision is close to `(0, -1)` destruction is triggered for the other entity.

The result is really cool because not only do entities get destroyed after a collision with the main character, but they also get destroyed when heavy collisions happen between each other.

https://user-images.githubusercontent.com/6058008/106210799-d2ebff80-61c7-11eb-88d0-957449f4a3d0.mp4

**Reply by encelo** [reply]
I should try to build it on Android. I'm eager to test it on my Nvidia Shield! 😉

**Reply by RealHippyChick** [reply]
This is a really cool project guys.. thanks for sharing! Super..

**Comment by Fahien** [comment]
Here is a new devlog with some new exciting stuff.

## Push and pull

Dynamic entities have been added to the game, and they can be pushed and pulled around the scene. This meant that new states should be added to characters, _push_ and _pull_ indeed.

The _push_ state is enabled when a character is trying to move and finds an obstacle either on its left or its right, quite easy to implement. Obstacles in the four directions (up, right, down, left) are collected after the physics system update by checking for physics body collisions and the normal of contact points. In addition, some [bitflags](https://en.wikipedia.org/wiki/Bit_field) are set which tell whether there are obstacles in any of the four directions.

https://user-images.githubusercontent.com/6058008/107680203-d38f9600-6c9d-11eb-8cc4-d157c80adc58.mp4

The _pull_ state is enabled when the player presses the `x` button and there is a dynamic obstacle, which is checked by querying the type of the obstacle's physics body. When the conditions are met, a [distance joint](https://box2d.org/documentation/md__d_1__git_hub_box2d_docs_dynamics.html#autotoc_md85) is created between the character and the dynamic entity to make sure that, when the character moves, the distance between the two bodies remains the same, effectively giving the ability to the character to carry around the entity. Of course, the joint is destroyed when exiting from the pull state.

## Data submodule

A [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) has been added to the repository under the `data/` directory, pointing to [Fahien/ncJump-data](https://github.com/Fahien/ncJump-data). There you could find the latest assets of the game: configurations, game maps, tile sets, and animations.

## A better editor

Many improvements have been added to the in-game editor: a  main menu bar at the top for choosing a placing mode between _tile_ and _entity_, changing the opacity of certain elements of the scene according to the current mode for an immediate visual feedback; a config window to tweak all sort of values not directly related to the gameplay, like the size of the window, the scale of the scene or the UI, the offset of the camera to move around the scene without necessarily moving the character.

https://user-images.githubusercontent.com/6058008/107680551-48fb6680-6c9e-11eb-9216-4c17db3469d8.mp4

## Conclusions

These updates can already be found on the [main branch](https://github.com/Fahien/ncJump/tree/main), but as an anticipation the more curious of you can try lurking into the [enemies branch](https://github.com/Fahien/ncJump/tree/enemies). 🍄

A big thanks to @encelo which is looking into the [project CI](https://github.com/Fahien/ncJump/tree/fix-ci) to make [things like this](https://ncine.github.io/ncjump/) possible (bug alert!).

**Comment by Fahien** [comment]
The time to add enemies to [the game](https://github.com/Fahien/ncJump) has finally come, and that translates to [factories](https://en.wikipedia.org/wiki/Factory_method_pattern)!

## Factories

An enemy is just another kind of entity, which is the reason that pushed me towards implementing a class which sole responsibility is to create entities: the `EntityFactory`. This factory contains a list of entity _prototypes_ which can be selected from the editor and placed into the scene. Under the hood, the prototype is cloned, the clone is moved at mouse position, and it is added to the list of entities of the scene.

https://user-images.githubusercontent.com/6058008/108392464-093efc80-7213-11eb-9955-8751a9b99b01.mp4

At this point I realized that every clone was duplicating textures, which is something that could and should be avoided, so I introduced another factory: the `GraphicsFactory`. The responsibility of this class is to maintain a list of textures which can be referenced by multiple _graphics components_, as well as to create any kind of sprite which makes use of one of those textures.

## Commands

Once I had enemies in the scene, I began wondering how to make them move. Reusing the state component of the main character sounded OK to me, but there was a problem. The movement logic within the state component was directly checking the gamepad and the keyboard, which meant that player's input would move both the main character and all the enemies. [Decoupling](https://en.wikipedia.org/wiki/Coupling_\(computer_programming\)) to the rescue.

```
Input → MoveCommand → StateComponent
```

I removed the dependency to the input from the state component, and I introduced the concept of commands. For simplicity, I will just focus on the `MoveCommand`. The state component now is just listening for move commands and, when they are issued, the component inspects the values in a move command and updates its state accordingly. Move commands, and possibly other kinds of commands, can be issued by any source: player input, in-game console, scripts, and so on.

## Scripts

Other engines may use terms like _behaviors_, but I prefer to stick with the technical term: _script_. With this you can indeed model behaviors of entities, like a wandering behavior that I had in mind for my first class of enemies. Ideally it might also be written with a scripting language, like [LUA](https://www.lua.org/) which is supported by the nCine, but that is something I will maybe explore in the future.

I wrote down a couple of line for the wandering behavior which should just move the entity to the right or to the left until an obstacle is encountered, and then it should turn back and continue in a loop. The result is quite interesting, but I believe it can be further improved. Bye for now!

https://user-images.githubusercontent.com/6058008/108392945-78b4ec00-7213-11eb-86b3-df4d988e8306.mp4

**Comment by Fahien** [comment]
It has been a while since my last devlog, but finally I landed it. It is full of good stuff so you would not regret the wait. Let us _ncJump_ into it!

## Dying state

Definitely one of the fundamental states. Any _living_ entity should die at some point, and this state represents that period of time where an animation should play telling the player an enemy is defeated or that the game is over. The [update method](https://gameprogrammingpatterns.com/update-method.html) of this state is super easy and it looks like that:

```cpp
void DyingState::update(Entity& entity)
{
    if (entity.animation.has_finished()) {
        entity.set_enabled(false);
        // just reuse this entity for something else
    }
}
```

https://user-images.githubusercontent.com/6058008/111708318-ee799b00-8845-11eb-91ce-63c76483d041.mp4

## Enabled flag

Of course this explains the `Entity::set_enabled()` method just introduced in the previous section. By calling it, we can effectively disable an entity and all its components without destroying it, or freeing its memory, so that it could be recycled to be reused later when we, for example, need to spawn another entity.

## Array of states

This was an important optimization. Before this, every state transitions triggered the destruction of the old state and creation of an instance of the new state, which is not really needed and does not scale nicely as a scene grows. Now, a state component keeps track of an array of all the possible state objects, together with a pointer telling us which one is the current state.

## Initial position

Another requirement for the scene is to define where the player should start. This _initial position_ is a pair of coordinates which tells us just this. When a scene _starts_, we put the player there.

## Definitions and factories

I love this pattern and I think I will use from now on for all of my game projects. Basically, we have to deal with big and heavy objects which are not easily serializable. Think of wrappers around third party components like Box2D bodies or nCine sprites. I tried to find a way to serialize those objects and I believe I found a very nice solution which is anyway something widely used, so I possibly just reinvented the wheel here, but that is ok.

I do not serialize those objects. I do serialize their definitions. A definition is a collection of all the parameters needed to construct an object. Let us take a _ncJump_ graphics component as an example.

```cpp
// Very simplified
auto def = GraphicsDef(GraphicsType::TILE);
def.path = String("image.png");
def.layer = 2;

auto gfx = GraphicsComponent(def);
```

As you can see, the graphics component constructor takes a graphics definition as a parameter. A graphics definition, like all other definitions, is a [Plain Old Data](https://en.wikipedia.org/wiki/Passive_data_structure) (POD) structure, which enables us to use [nlohmann macros](https://github.com/nlohmann/json#simplify-your-life-with-macros) to quickly define all the functions needed to serialize it. Done. Awesome.

```cpp
// That's it
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(GraphicsDef, type, path, layer);
```

## CI and automatic deploy

This will blown your mind, because the project now builds automatically, through GitHub actions, for [Linux, Window, MacOS, Android](https://github.com/Fahien/ncJump-artifacts), and believe it or not for the web. Indeed, [click here](https://www.antoniocaggiano.eu/ncJump-artifacts/) and be prepared to say **wow**. All the credits for this go to the brilliant nCine author: @encelo!

**Comment by encelo** [comment]
ncJump now runs (and jumps) on the Steam Deck too! 💪🏻 

[![ncJump running on the Steam Deck](https://img.youtube.com/vi/ZKdDtJiIUdo/0.jpg)](https://www.youtube.com/watch?v=ZKdDtJiIUdo)
Click the thumbnail to open the YouTube video.
